src/rules/fileNameCasingRule.ts

```diff
@@ -28,6 +28,74 @@ enum Casing {
     SnakeCase = "snake-case",
 }
 
+type RegexConfig = Record<string, Casing>;
+
+type SimpleConfig = Casing;
+
+type Config = SimpleConfig | RegexConfig;
+
+interface ValidatorSuccessfulResult {
+    valid: true;
+}
+
+interface ValidatorFailureResult {
+    valid: false;
+    failedCasing: Casing;
+}
+
+type ValidatorResult = ValidatorSuccessfulResult | ValidatorFailureResult;
+
+type Validator<T> = (casing: T) => (sourceFile: ts.SourceFile) => ValidatorResult;
+
+const rules = [Casing.CamelCase, Casing.PascalCase, Casing.KebabCase, Casing.SnakeCase];
+
+function isCorrectCasing(fileName: string, casing: Casing): boolean {
+    switch (casing) {
+        case Casing.CamelCase:
+            return isCamelCased(fileName);
+        case Casing.PascalCase:
+            return isPascalCased(fileName);
+        case Casing.KebabCase:
+            return isKebabCased(fileName);
+        case Casing.SnakeCase:
+            return isSnakeCased(fileName);
+    }
+}
+
+const RegexValidator: Validator<RegexConfig> = casingConfig => (sourceFile): ValidatorResult => {
+    const fileName = path.parse(sourceFile.fileName).base;
+    const config = Object.keys(casingConfig).map(key => ({
+        casing: casingConfig[key],
+        regex: RegExp(`${key}$`),
+    }));
+
+    const match = config.find(c => c.regex.test(fileName));
+
+    if (match === undefined) {
+        return { valid: true };
+    }
+
+    const normalizedFileName = fileName.replace(match.regex, "");
+
+    return isCorrectCasing(normalizedFileName, match.casing)
+        ? { valid: true }
+        : { valid: false, failedCasing: match.casing };
+};
+
+const SimpleValidator: Validator<Casing> = casingConfig => (sourceFile): ValidatorResult => {
+    const fileName = path.parse(sourceFile.fileName).name;
+    const isValid = isCorrectCasing(fileName, casingConfig);
+
+    return isValid ? { valid: true } : { valid: false, failedCasing: casingConfig };
+};
+
+const validate = (sourceFile: ts.SourceFile, casingConfig: Config): ValidatorResult => {
+    const validator = typeof casingConfig === "string" ? SimpleValidator : RegexValidator;
+
+    // @ts-ignore https://github.com/Microsoft/TypeScript/issues/7294
```

Comment:
Don't break my beautiful lambda 😭 

Yeah I did it this way because on testing (or usage) I could apply the same `Config` to different files but yeah, too fancy without needs I guess 😅